mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala

```diff
@@ -0,0 +1,690 @@
+ * the root. Then, it picks up the larger child of the two and merge it to the cluster tree.
+ * After that, there are tree candidates to merge. Those are the smaller child of the root and
+ * the two children of the larger cluster of the root. It picks up the largest cluster of the tree
+ * and merge it to the * cluster tree. Like this, it continues to pick up the largest one of the
+ * candidates and merge it to the cluster tree until the desired number of clusters is reached.
+ *
+ * @param k tne desired number of clusters
+ * @param clusterMap the pairs of cluster and its index as Map
+ * @param maxIterations the number of maximal iterations to split clusters
+ * @param seed a random seed
+ */
+@Since("1.6.0")
+class BisectingKMeans private (
+    private var k: Int,
+    private var clusterMap: Map[BigInt, BisectingClusterNode],
+    private var maxIterations: Int,
+    private var seed: Long) extends Logging {
+
+  /**
+   * Constructs with the default configuration
+   */
+  @Since("1.6.0")
+  def this() = this(20, mutable.ListMap.empty[BigInt, BisectingClusterNode], 20, 1)
+
+  /**
+   * Sets the number of clusters you want
+   */
+  @Since("1.6.0")
+  def setK(k: Int): this.type = {
+    this.k = k
+    this
+  }
+
+  @Since("1.6.0")
+  def getK: Int = this.k
+
+  /**
+   * Sets the number of maximal iterations in each clustering step
+   */
+  @Since("1.6.0")
+  def setMaxIterations(maxIterations: Int): this.type = {
+    this.maxIterations = maxIterations
+    this
+  }
+
+  @Since("1.6.0")
+  def getMaxIterations: Int = this.maxIterations
+
+  /**
+   * Sets the random seed
+   */
+  @Since("1.6.0")
+  def setSeed(seed: Long): this.type = {
+    this.seed = seed
+    this
+  }
+
+  @Since("1.6.0")
+  def getSeed: Long = this.seed
+
+  /**
+   * Runs the bisecting k-means algorithm
+   * @param input RDD of vectors
+   * @return model for the bisecting kmeans
+   */
+  @Since("1.6.0")
+  def run(input: RDD[Vector]): BisectingKMeansModel = {
+    val sc = input.sparkContext
+
+    // `clusterStats` is described as binary tree structure
+    // `clusterStats(1)` means the root of a binary tree
+    var clusterStats = mutable.Map.empty[BigInt, BisectingClusterStat]
+    var step = 1
+    var noMoreDividable = false
+    var rddArray = Array.empty[RDD[(BigInt, BV[Double])]]
+    // the number of maximum nodes of a binary tree by given parameter
+    val multiplier = math.ceil(math.log10(this.k) / math.log10(2.0)) + 1
+    val maxAllNodesInTree = math.pow(2, multiplier).toInt
+
+    // divide clusters until the number of clusters reachs the condition
+    // or there is no dividable cluster
+    val startTime = System.currentTimeMillis()
+    var data = BisectingKMeans.initData(input).cache()
+    while (clusterStats.size < maxAllNodesInTree && noMoreDividable == false) {
+      logInfo(s"${sc.appName} starts step ${step}")
+      val leafClusters = BisectingKMeans.summarizeClusters(data)
```

Comment:
* `leafCluster` -> `leafClusterStats`
* add a TODO to remove non-leaf cluster stats from the broadcast data